### PR TITLE
EVG-5857: add diff to commit queue patches

### PR DESCRIPTION
**Pull request by ybrill**
Add the diff GiHub provides for a PR to the patch so it can be displayed in the UI.

For example:
Screen Shot 2019-03-11 at 1 39 32 PM

### CODE REVIEW

File: thirdparty/github.go [hunk 1]
```diff
@@ -727,7 +727,7 @@ func doGithubRequest(client *http.Client, req *http.Request, accept string) (str
 
 // buildPatchURL creates a URL to enable downloading patch files through the
 // Github API
-func buildPatchURL(gp *patch.GithubPatch) string {
+func buildPatchURL(gp patch.GithubPatch) string {
```
Comment (ybrill) [inline thirdparty/github.go]: This was expecting a pointer for some reason.
Many of the changes in the PR are just to pass `gp` by value.

File: model/patch/patch.go [hunk 2]
```diff
@@ -370,6 +370,9 @@ func MakeMergePatch(pr *github.PullRequest, projectID, alias string) (*Patch, er
 		GithubPatchData: GithubPatch{
 			PRNumber:       pr.GetNumber(),
 			MergeCommitSHA: pr.GetMergeCommitSHA(),
+			BaseOwner:      pr.Base.User.GetLogin(),
```
Comment (ybrill) [inline model/patch/patch.go]: This information is used later to get the diff from GitHub

File: units/commit_queue.go [hunk 3]
```diff
@@ -321,6 +320,28 @@ func makeMergePatch(ctx context.Context, pr *github.PullRequest, githubToken, pr
 		return nil, nil, errors.Wrap(err, "can't make patch")
 	}
 
+	// diff to be displayed in the UI
+	patchContent, summaries, err := thirdparty.GetGithubPullRequestDiff(ctx, githubToken, patchDoc.GithubPatchData)
```
Comment (ybrill) [inline units/commit_queue.go]: This is the main change.

File: units/commit_queue.go [hunk 4]
```diff
@@ -253,15 +255,12 @@ func getModules(ctx context.Context, githubToken string, nextItem *commitqueue.C
 			return nil, nil, true, errors.Wrapf(err, "module '%s' misconfigured (malformed URL)", mod.Module)
 		}
 
-		githash := ""
-		if mod.Issue != "" {
```
Comment (ybrill) [inline units/commit_queue.go]: This is never the case (it's required by the regex used to parse the modules) and doesn't make sense anyway (what's a patch without a place to get it from?)

File: units/commit_queue.go [hunk 5]
```diff
@@ -321,6 +320,28 @@ func makeMergePatch(ctx context.Context, pr *github.PullRequest, githubToken, pr
...
+	}
+
+	patchFileID := fmt.Sprintf("%s_%s", patchDoc.Id.Hex(), patchDoc.Githash)
+	if err = db.WriteGridFile(patch.GridFSPrefix, patchFileID, strings.NewReader(patchContent)); err != nil {
+		return nil, nil, errors.Wrap(err, "failed to write patch file to db")
+	}
+
+	// empty string for name of the main patch
+	patchDoc.Patches = append(patchDoc.Patches, patch.ModulePatch{
+		ModuleName: "",
```
Comment (tychoish) [inline units/commit_queue.go]: nit: this is the default value, we don't need to specify it here.

File: units/commit_queue.go [hunk 6]
```diff
@@ -284,10 +283,10 @@ func validatePR(pr *github.PullRequest) error {
 	if pr.GetMergeCommitSHA() == "" {
 		catcher.Add(errors.New("no merge commit SHA"))
 	}
-	if pr.GetUser() == nil || pr.GetUser().GetID() == 0 {
+	if pr.GetUser() == nil || pr.GetUser().GetLogin() == "" {
 		catcher.Add(errors.New("no valid user"))
 	}
-	if pr.GetBase() == nil || pr.GetBase().GetSHA() == "" {
+	if pr.GetBase() == nil || pr.GetBase().GetSHA() == "" || pr.GetBase().GetRef() == "" {
```
Comment (tychoish) [inline units/commit_queue.go]: does it make sense to make this a method on the pr type?

Comment (ybrill) [reply units/commit_queue.go]: The pr type is from go-github

Comment (tychoish) [reply units/commit_queue.go]: a helper function to reduce the complexity of the if clause then would help clean things up a bit and make the code a bit more self documenting

Comment (ybrill) [reply units/commit_queue.go]: done. I also relocated it to the thirdparty package.

File: units/commit_queue.go (hunk 5 above)
Comment (ybrill) [reply units/commit_queue.go]: done.

**Comment by ybrill** [comment]
> This looks reasonable, but is there a way to unit test it?

@bsamek: refactored to add testing

**Pull request merged**